implementations/rust/ockam/ockam/src/system/hooks/delivery.rs

```diff
@@ -0,0 +1,170 @@
+use crate::{Context, DelayedEvent, OckamError, OckamMessage, Result, Routed, SystemHandler};
+use ockam_core::{
+    async_trait,
+    compat::{boxed::Box, collections::BTreeMap, string::String, vec::Vec},
+    Address, Any, Decodable, Encodable,
+};
+
+#[derive(Clone, Default)]
+pub struct SenderConfirm {
+    next: Option<Address>,
+    journal: BTreeMap<Address, OckamMessage>,
+}
+
+fn to_str(v: &[u8]) -> &str {
+    core::str::from_utf8(v).unwrap()
```

Comment:
We probably should handle this better, given that we don't require metadata values be UTF-8 (or strings at all)...

Reply:
I don't really see the issue with this considering it's an internal implementation detail to the protocol between different pipe worker endpoints. In this case we _do_ enforce UTF-8 but this detail isn't exposed to the user at all.